File: solutions/java/resistor-color-trio/1/src/main/java/ResistorColorTrio.java
import java.util.List;

class ResistorColorTrio {

    private final String LABEL_FORMAT = "%.0f %sohms";

    String label(String[] colors) {
        int base = getBase(colors);
        double magnitude = getMagnitude(colors);

        double value = base * magnitude;

        String prefix = "";

        if (value == 0) {
            return String.format(LABEL_FORMAT, value, prefix);
        }

        if (value % 1_000_000_000 == 0) {
            prefix = "giga";
            value = value / 1_000_000_000;
        } else if (value % 1_000_000 == 0) {
            prefix = "mega";
            value = value / 1_000_000;
        } else if (value % 1_000 == 0) {
            prefix = "kilo";
            value = value / 1_000;
        }

        return String.format(LABEL_FORMAT, value, prefix);
    }

    double getMagnitude(String[] colors) {
        String color = colors[2];

        int numZeros = colorCode(color);

        return Math.pow(10, numZeros);
    }

    int getBase(String[] colors) {
        StringBuilder resistorValue = new StringBuilder();

        int count = 0;

        for (String color : colors) {
            count++;
            if (count > 2) {
                break;
            }
            int value = colorCode(color);

            if (value != -1) {
                resistorValue.append(value);
            }
        }

        return Integer.parseInt(resistorValue.toString(), 10);
    }

    private int colorCode(String color) {
        return colors().indexOf(color);
    }

    private List<String> colors() {
        return List.of(
                "black",
                "brown",
                "red",
                "orange",
                "yellow",
                "green",
                "blue",
                "violet",
                "grey",
                "white");
    }
}

File: solutions/csharp/resistor-color-trio/1/ResistorColorTrio.cs
public static class ResistorColorTrio
{
    
    private const int NumberOfSignicantFigures = 2;
    
    public static string Label(string[] colors)
    {
        if (colors.Length < 3)
        {
            throw new ArgumentException("Not enough bands to determine resistance", nameof(colors) );
        }
        
        var resistance = GetResistance(colors);

        var formatted = FormattingWithPrefix(resistance);

        return $"{formatted.resistance} {formatted.prefix}ohms";
    }
    
    private static long GetResistance(string[] colors)
    {
        var firstTwo = colors
            .Take(NumberOfSignicantFigures)
            .Select((color, index) =>
            {
                var multiplier = (long)Math.Pow(10, NumberOfSignicantFigures - index - 1);
                var resistance = ColorToResistance(color);
                
                return multiplier * resistance;
            })
            .Sum();

        var multiplier = (long)Math.Pow(10, ColorToResistance(colors[2]));
        
        return firstTwo * multiplier;
    }
    
    private static (long resistance, string prefix) FormattingWithPrefix(long resistance) => resistance switch
    {
        0 => (resistance, ""),
        _ when resistance % 1_000_000_000 == 0 => (resistance / 1_000_000_000, "giga"),
        _ when resistance % 1_000_000 == 0 => (resistance / 1_000_000, "mega"),
        _ when resistance % 1_000 == 0 => (resistance / 1_000, "kilo"),
        _ => (resistance, "")
    };

    
    private static int ColorToResistance(string color) => color.ToLower() switch
    {
        "black" => 0,
        "brown" => 1,
        "red" => 2,
        "orange" => 3,
        "yellow" => 4,
        "green" => 5,
        "blue" => 6,
        "violet" => 7,
        "grey" => 8,
        "white" => 9,
        _ => throw new ArgumentOutOfRangeException(nameof(color), color, null)
    };
}
